src/app/features/worklog/worklog-export/worklog-export.util.ts

```diff
@@ -0,0 +1,274 @@
+import * as moment from 'moment';
+import { msToClockString } from '../../../ui/duration/ms-to-clock-string.pipe';
+import { msToString } from '../../../ui/duration/ms-to-string.pipe';
+import { roundDuration } from '../../../util/round-duration';
+import { roundTime } from '../../../util/round-time';
+import { unique } from '../../../util/unique';
+import { Project, ProjectCopy } from '../../project/project.model';
+import { Tag, TagCopy } from '../../tag/tag.model';
+import { WorklogTask } from '../../tasks/task.model';
+import { WorkStartEnd } from '../../work-context/work-context.model';
+import { WorklogExportSettingsCopy, WorklogGrouping } from '../worklog.model';
+
+const LINE_SEPARATOR = '\n';
+const EMPTY_VAL = ' - ';
+
+interface RowItem {
+  dates: string[];
+  workStart: number | undefined;
+  workEnd: number | undefined;
+  timeSpent: number;
+  timeEstimate: number;
+  tasks: WorklogTask[];
+  titles?: string[];
+  titlesWithSub?: string[];
+  notes: string[];
+  projects: string[];
+  tags: string[];
+}
+
+export const createRows = (tasks: WorklogTask[], startTimes: WorkStartEnd, endTimes: WorkStartEnd,
+  groupBy: WorklogGrouping, allProjects: Project[], allTags: Tag[]): RowItem[] => {
+
+  const mapToGroups = (task: WorklogTask) => {
+    const taskGroups: { [key: string]: RowItem } = {};
+    const createEmptyGroup = (): RowItem => {
+      return {
+        dates: [],
+        timeSpent: 0,
+        timeEstimate: 0,
+        tasks: [],
+        titlesWithSub: [],
+        titles: [],
+        notes: [],
+        projects: [],
+        tags: [],
+        workStart: undefined,
+        workEnd: undefined,
+      };
+    };
+
+    // If we're grouping by parent task ignore subtasks
+    // If we're grouping by task ignore parent tasks
+    if ((groupBy === WorklogGrouping.PARENT && task.parentId !== null)
+      || (groupBy === WorklogGrouping.TASK && task.subTaskIds.length > 0)
+    ) {
+      return taskGroups;
+    }
+
+    switch (groupBy) {
+      case WorklogGrouping.DATE:
+        if (!task.timeSpentOnDay) {
+          return {};
+        }
+        const numDays = Object.keys(task.timeSpentOnDay).length;
+        Object.keys(task.timeSpentOnDay).forEach(day => {
+          taskGroups[day] = {
+            dates: [day],
+            tasks: [task],
+            notes: (task.notes) ? [task.notes] : [],
+            projects: (task.projectId) ? [task.projectId] : [],
+            tags: task.tagIds,
+            workStart: startTimes[day],
+            workEnd: endTimes[day],
+            timeSpent: 0,
+            timeEstimate: 0
+          };
+          if (!task.subTaskIds || task.subTaskIds.length === 0) {
+            taskGroups[day].timeSpent = task.timeSpentOnDay[day];
+            taskGroups[day].timeEstimate = task.timeEstimate / numDays;
+          }
+        });
+        break;
```

Comment:
Maybe using 
`return workGroups;`

Makes it easier to grasp what is happening.